python/cudf/cudf/core/dataframe.py

```diff
@@ -260,11 +270,89 @@ def __init__(self, data=None, index=None, columns=None, dtype=None):
 
+    def _get_union_of_indices(self, indexes):
+        if len(indexes) == 1:
+            return indexes[0]
+        else:
+            merged_index = cudf.core.Index._concat(indexes)
+            merged_index = merged_index.drop_duplicates()
+            _, inds = merged_index._values.sort_by_values()
+            return merged_index.take(inds)
+
+    def _get_union_of_series_names(self, series_list):
+        names_list = []
+        unnamed_count = 0
+        for idx, series in enumerate(series_list):
+            if series.name is None:
+                names_list.append(f"Unnamed {unnamed_count}")
+                unnamed_count += 1
+            else:
+                names_list.append(series.name)
+        if unnamed_count == len(series_list):
+            names_list = [*range(len(series_list))]
+
+        return names_list
+
+    def _init_from_series_list(self, data, columns, index):
+        dataframe_columns = self._get_union_of_indices([d.index for d in data])
+
+        if index is None:
+            index = as_index(self._get_union_of_series_names(data))
+        else:
+            data_length = len(data)
+            index_length = len(index)
+            if data_length != index_length:
+                if index_length % data_length == 0:
+                    initial_data = data
+                    data = []
+                    for _ in range(int(index_length / data_length)):
+                        data.extend([o.copy() for o in initial_data])
+                else:
+                    raise ValueError(
+                        f"Shape of passed values is "
+                        f"{(data_length, len(data[0]))}, "
+                        f"indices imply {(index_length, len(data[0]))}"
+                    )
+
+            index = as_index(index)
+
+        series_lengths = list(map(lambda x: len(x), data))
+        data = numeric_normalize_types(*data)
+        if series_lengths.count(series_lengths[0]) == len(series_lengths):
+            for idx, series in enumerate(data):
+                if not series.index.equals(dataframe_columns):
+                    series = series.reindex(dataframe_columns)
+                self._data[idx] = column.as_column(series._column)
+
+            self._index = dataframe_columns
+
+            transpose = self.T
+
+        else:
+            concat_df = cudf.concat(data, axis=1)
+
+            if concat_df.columns.dtype == "object":
+                concat_df.columns = concat_df.columns.astype("str")
+
+            transpose = concat_df.T
+
+        transpose._index = index
+        self._mimic_inplace(transpose, inplace=True)
+
+        if columns:
+            for col_name in columns:
+                if col_name not in self._data:
+                    self._data[col_name] = column.column_empty(
+                        row_count=len(self), dtype=None, masked=True
+                    )
+            self._data = self._data.get_by_label(columns)
```

Comment:
Struggled to fully wrap my head around what's going on in this chunk of code, will connect offline to have a quick call to walk me through this

Reply:
Resolving this as I have added some comments so that other reviewers can also get an understanding of the code. But anyone who is reviewing this part of code, please feel free to connect offline.